# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

- **Consider a significance level of 5% for all tests.**

In [73]:
# import numpy and pandas

import pandas as pd
import numpy as np
from scipy import stats as st

# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

To load the data, run the code below.

In [61]:
# Run this code:

pokemon = pd.read_csv('../pokemon.csv')

Let's start off by looking at the `head` function in the cell below.

In [63]:
# Your code here:

pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [65]:
# Your code here:

pokemon.Legendary.value_counts()

Legendary
False    735
True      65
Name: count, dtype: int64

Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [67]:
# Subset wit non/legendary
legendary = pokemon.loc[pokemon["Legendary"] == True]
non_legendary = pokemon.loc[pokemon["Legendary"] == False]

# mean and std of Total for legendary Pokemon
print(f"The mean for the column 'Total' for legendary Pokemons is: {legendary.Total.mean().round(2)}.")
print(f"The std for the column 'Total' for legendary Pokemons is: {round(legendary.Total.std(), 2)}.")
print("")

# mean and std of Total for non-legendary Pokemon
print(f"The mean for the column 'Total' for non-legendary Pokemons is: {non_legendary.Total.mean().round(2)}.")
print(f"The std for the column 'Total' for non-legendary Pokemons is: {round(non_legendary.Total.std(), 2)}.")

The mean for the column 'Total' for legendary Pokemons is: 637.38.
The std for the column 'Total' for legendary Pokemons is: 60.94.

The mean for the column 'Total' for non-legendary Pokemons is: 417.21.
The std for the column 'Total' for non-legendary Pokemons is: 106.76.


The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

In [111]:
# Your code here:
points_legendary = legendary["Total"]
points_non_legendary = non_legendary["Total"]

t_stat, p_value = st.ttest_ind(points_legendary, points_non_legendary, equal_var=False)
print(f"Test Statistic (t): {t_stat:.2f}")
print(f"P-value: {p_value:.30f}")

Test Statistic (t): 19.06
P-value: 0.000000000000000000000000000021


What do you conclude from this test? Write your conclusions below.

# Your conclusions here:
The difference in the number of total points between the two groups seems to be significant.

How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [123]:
# Your code here:

pokemon["Type 1"].value_counts().sort_values(ascending=False)

Type 1
Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Ghost        32
Ground       32
Dragon       32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: count, dtype: int64

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [165]:
# Your code here:

# Subset Type 1 Water/ Others
water = pokemon.loc[pokemon["Type 1"] == "Water"]
non_water = pokemon.loc[pokemon["Type 1"] != "Water"]

mean_water = round(water["Total"].mean(), 2)
std_water = round(water["Total"].std(), 2)

mean_non_water = round(non_water["Total"].mean(), 2)
std_non_water = round(non_water["Total"].std(), 2)

print(f"The mean for the total points of Pokemons whose Type 1 is 'Water' is: {mean_water} and the std is: {std_water}.")
print(f"The mean for the total points for the other Pokemons is: {mean_non_water} and the std is: {std_non_water}.")

The mean for the total points of Pokemons whose Type 1 is 'Water' is: 430.46 and the std is: 113.19.
The mean for the total points for the other Pokemons is: 435.86 and the std is: 121.09.


Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

# hypothesis
- Null hypothesis: The means of the two groups are not significantly different.
- Alternative hypothesis: The means of the two groups differ significantly.

In [169]:
# t-test

points_water = water["Total"]
points_non_water= non_water["Total"]

t_stat, p_value = st.ttest_ind(points_water, points_non_water, equal_var=True)
print(f"Test Statistic (t): {t_stat:.2f}")
print(f"P-value: {p_value:.2f}")

Test Statistic (t): -0.44
P-value: 0.66


Write your conclusion below.

# Your conclusions here:

- The p-value is much greater than alpha (0.05), so we have no grounds to reject the Null Hypothesis.
- The means of the two groups regarding their total points are not significantly different.

# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

In [171]:
# Your code here:

from scipy.stats import ttest_rel

In [173]:
pokemon.head(1)

#       Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0  1  Bulbasaur  Grass  Poison    318  45      49       49       65       65   

   Speed  Generation  Legendary  
0     45           1      False

Describe the results of the test in the cell below.

In [179]:
# Your conclusions here:

t_stat, p_value = st.ttest_rel(pokemon.Attack, pokemon.Defense)
print(f"Test Statistic (t): {t_stat:.2f}")
print(f"P-value: {p_value:.6f}")

Test Statistic (t): 4.33
P-value: 0.000017


We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

In [185]:
# Your code here:

t_stat, p_value = st.ttest_rel(pokemon["Sp. Atk"], pokemon["Sp. Def"])
print(f"Test Statistic (t): {t_stat:.2f}")
print(f"P-value: {p_value:.2f}")

Test Statistic (t): 0.85
P-value: 0.39


Describe the results of the test in the cell below.

# Your conclusions here:

- The low p-value regarding the scores in Attack and Defense show that there is a significant difference betwenn those two, so the hypothesis (that the defense and attack scores are equal) should be rejected.
- This is not the case for the Special Attack and the Special Defense. These two values do not differ significantly and the null hypothesis cannot be rejected here.

As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

In [190]:
from scipy.stats import ttest_1samp

- Null hypothesis: the difference between the means of defense and attack scores is zero
- Alternative hypothesis: the difference between the means of defense and attack scores is not zero

In [202]:
# Your code here:

mu = pokemon.Attack.mean()

t_stat, p_value = st.ttest_1samp(pokemon.Defense, mu)
print(f"Test Statistic (t): {t_stat:.2f}")
print(f"P-value: {p_value:.7f}")

Test Statistic (t): -4.68
P-value: 0.0000034


***Comment***
I did not understand how to do this. I checked the parameter of ttest_1samp and decided to put in one column as data and the mean of the other as mu. But the results are not the same with the ttest_rel test.

# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

In [216]:
# Your code here:

crosstab_result = pd.crosstab(index=pokemon["Legendary"], columns=pokemon["Type 1"] == "Water")
crosstab_result

Type 1     False  True 
Legendary              
False        627    108
True          61      4

Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [210]:
from scipy.stats import chi2_contingency

In [222]:
# Your code here:

chi2_stats, chi2_pvalue, _, _ = chi2_contingency(crosstab_result)
chi2_pvalue.round(4)

0.0863

Based on a 95% confidence, should we reject the null hypothesis?

# Your answer here:

Yes, we should reject the null hypothesis. With a p-value of 0.086 which is greater than an alpha of 0.05, there is not enough evidence that there is a significant reationship between the two.